freqtrade/freqai/freqai_interface.py

```diff
@@ -354,25 +357,43 @@ def start_backtesting(
                             f"Message: {msg}, skipping.", exc_info=True)
+
+                        if self.new_feature_selector:
+                            logger.warning(
+                                "Train failed. Try to train on next pair." \
+                                    if self.new_feature_selector else
+                                "Train failed. Raise error, fix data issue and try again."
+                            )
+                            self.tb_logger.close()
                         self.model = None
 
-                    self.dd.pair_dict[pair]["trained_timestamp"] = int(
-                        tr_train.stopts)
-                    if self.plot_features and self.model is not None:
-                        plot_feature_importance(self.model, pair, dk, self.plot_features)
-                    if self.save_backtest_models and self.model is not None:
-                        logger.info('Saving backtest model to disk.')
-                        self.dd.save_data(self.model, pair, dk)
-                    else:
-                        logger.info('Saving metadata to disk.')
-                        self.dd.save_metadata(dk)
+                    hard_check_model_valid = True
+                    if self.new_feature_selector:
+                        hard_check_model_valid = bool(False if self.model is None else True)
+                        if hard_check_model_valid:
+                            logger.info(
+                                "Model is trained. Saving model and metadata to disk."
+                            )
+
+                    if hard_check_model_valid:
+                        self.dd.pair_dict[pair]["trained_timestamp"] = int(tr_train.stopts)
+                        if self.plot_features and self.model is not None:
+                            plot_feature_importance(self.model, pair, dk, self.plot_features)
+                        if self.save_backtest_models and self.model is not None:
+                            logger.info('Saving backtest model to disk.')
+                            self.dd.save_data(self.model, pair, dk)
+                        else:
+                            logger.info('Saving metadata to disk.')
+                            self.dd.save_metadata(dk)
+
                 else:
                     self.model = self.dd.load_data(pair, dk)
 
-                pred_df, do_preds = self.predict(dataframe_backtest, dk)
-                append_df = dk.get_predictions_to_append(pred_df, do_preds, dataframe_backtest)
-                dk.append_predictions(append_df)
-                dk.save_backtesting_prediction(append_df)
+                if hard_check_model_valid:
+                    pred_df, do_preds = self.predict(dataframe_backtest, dk)
+                    append_df = dk.get_predictions_to_append(pred_df, do_preds, dataframe_backtest)
+                    dk.append_predictions(append_df)
```

Comment:
The return dataframe (the DF the leaves `start_backtesting`) needs to match the length of the incoming dataframe length (the DF that enters `start_backtesting`). If I understand your goal here is to simply skip appending anything to the return dataframe. This would result in negative consequences back in the strategy. 

So to properly handle this, you would need some functionality to append some NaN or zeross in place of the desired predictions. We also need to ensure that this specific failed range has `do_predict=0` so that the strategy won't try to act on the 0s that we are filling here.

We need to keep in mind that the failure of training could be due to a lack of data *or* some other failure in the training. It would be good to distinguish these and log them so the user knows why things are failing but continuing. We can do a check for data length before even trying to train, if it is 0, then we can flag right there that there is no data available for training for this particular time-range. Otherwise, we can try the train and catch exception to report an issue that occurs in a particular training.

Please follow up with any questions or clarifications you have regarding this points, to avoid confusion.

Reply:
Do I understand correctly that you mean the following conversions:
- Adding NaN (Not a Number) in the place of expected predictions if the length of the returned dataframe is less than the length of the incoming dataframe.
- Adding zeros (0) to the place of expected predictions if the length of the returned dataframe is less than the length of the incoming dataframe.
- Setting the do_predict parameter to 0 for a failed range so that the strategy does not attempt to act on filled zeros.

The most important thing is to add NaN or zeros to the dataframe so that it has a length equal to the length of the incoming dataframe and zeros to distinguish between empty data and mark the training dataframe set to 0 if there are no other predictions? I could make multiple options for processing the dataset, not just true or false.

Reply:
Have a look at my explanation of the `start_backtesting()` procedure. 

This is not about checking if the length of the returned dataframe is less than the length of the incoming dataframe. 

This is about ensuring that for each iteration of a sub-timerange, you always append the same amount of candles, whether those candles are real predictions from a properly trained model, or filler predictions from a failure. Same number appended, then moving to next sub-timerange.

Reply:
### What Needs to Be Done:
- Match the lengths of dataframes: Ensure that the length of the return dataframe matches the length of the incoming dataframe. This can be achieved by adding NaN or zeros in place of expected predictions if the number of predictions is less than required.
- Handle failures: If there is a failure during training, such as due to a lack of data or other issues, it should be handled appropriately.
- Set do_predict=0 for failed ranges so that the strategy does not attempt to act on filled zeros.
- Log failures so that users are aware of the reason for the training failure.
- Distinguish between types of failures: Differentiate between different types of failures (e.g., lack of data versus other issues) and log them accordingly.
- Check data length before training: Before attempting to train the model, check the length of the incoming data. If there is no data, immediately indicate the issue and avoid training attempts.
- Consistency and robustness: Ensure consistency in the process across different iterations and handle data anomalies appropriately.
### What Should Not Be Done:
- Do not ignore errors: Ignoring errors during the backtesting process can lead to inaccurate results and incorrect decision-making.
- Do not skip error logging: Being unaware of what happens during training or testing can make diagnosing and fixing problems in the future difficult.
- Do not alter the length of the incoming dataframe: The length of the incoming dataframe should not be changed; all manipulations should be done only with the return dataframe.
- Do not overdo detailing: Although detailing is important for understanding the process, excessive detailing can complicate implementation.

Did I understand correctly?